From 63b187e3861ebc0111649c5c8e669a718f611051 Mon Sep 17 00:00:00 2001
From: Simon Xue <xxm@rock-chips.com>
Date: Thu, 10 Oct 2019 15:15:52 +0800
Subject: [PATCH] PCI: rockchip: add ref_count for
 rk_pcie_misc_open/rk_pcie_misc_release

Don't rechanged the params of dma_trx_obj if /dev/pcie-dev be opened more
than one time

Change-Id: Ie73870e12115d091dcd33afa604c8256388b82b9
Signed-off-by: Simon Xue <xxm@rock-chips.com>
---
 drivers/pci/controller/rockchip-pcie-dma.c | 16 ++++++++++++++++
 drivers/pci/controller/rockchip-pcie-dma.h |  2 ++
 2 files changed, 18 insertions(+)

diff --git a/drivers/pci/controller/rockchip-pcie-dma.c b/drivers/pci/controller/rockchip-pcie-dma.c
index e44d451ec88a..41438ab89ab1 100644
--- a/drivers/pci/controller/rockchip-pcie-dma.c
+++ b/drivers/pci/controller/rockchip-pcie-dma.c
@@ -362,6 +362,10 @@ static int rk_pcie_misc_open(struct inode *inode, struct file *filp)
 
 	filp->private_data = pcie_misc_dev->obj;
 
+	mutex_lock(&pcie_misc_dev->obj->count_mutex);
+	if (pcie_misc_dev->obj->ref_count++)
+		goto already_opened;
+
 	pcie_misc_dev->obj->loop_count = 0;
 	pcie_misc_dev->obj->local_read_available = 0x0;
 	pcie_misc_dev->obj->local_write_available = 0xff;
@@ -370,13 +374,23 @@ static int rk_pcie_misc_open(struct inode *inode, struct file *filp)
 
 	pr_info("Open pcie misc device success\n");
 
+already_opened:
+	mutex_unlock(&pcie_misc_dev->obj->count_mutex);
 	return 0;
 }
 
 static int rk_pcie_misc_release(struct inode *inode, struct file *filp)
 {
+	struct dma_trx_obj *obj = filp->private_data;
+
+	mutex_lock(&obj->count_mutex);
+	if (--obj->ref_count)
+		goto still_opened;
+
 	pr_info("Close pcie misc device\n");
 
+still_opened:
+	mutex_unlock(&obj->count_mutex);
 	return 0;
 }
 
@@ -715,8 +729,10 @@ struct dma_trx_obj *rk_pcie_dma_obj_probe(struct device *dev)
 
 	obj->irq_num = 0;
 	obj->loop_count_threshold = 0;
+	obj->ref_count = 0;
 	init_completion(&obj->done);
 
+	mutex_init(&obj->count_mutex);
 	rk_pcie_add_misc(obj);
 
 #ifdef CONFIG_DEBUG_FS
diff --git a/drivers/pci/controller/rockchip-pcie-dma.h b/drivers/pci/controller/rockchip-pcie-dma.h
index 57b884f0b7ee..90888328fd01 100644
--- a/drivers/pci/controller/rockchip-pcie-dma.h
+++ b/drivers/pci/controller/rockchip-pcie-dma.h
@@ -170,6 +170,8 @@ struct dma_trx_obj {
 	int				busno;
 	void				*priv;
 	struct completion		done;
+	int				ref_count;
+	struct mutex			count_mutex;
 	unsigned long			irq_num;
 	struct dentry			*pcie_root;
 };
-- 
2.35.3

